src/main/java/com/emc/object/s3/S3SignerV4.java

```diff
@@ -0,0 +1,397 @@
+        }
+
+        return canonicalizedHeaders;
+    }
+
+    protected String getStringToSign(String method, String resource, Map<String, String> parameters,
+                                     Map<String, List<Object>> headers, String date, String service, String canonicalRequest) {
+        StringBuilder stringToSign = new StringBuilder();
+        stringToSign.append(S3Constants.AWS_HMAC_SHA256_ALGORITHM).append("\n");
+        stringToSign.append(date).append("\n");
+        SimpleDateFormat sdf = new SimpleDateFormat(AMZ_DATE_FORMAT, Locale.US);
+        try {
+            Date d = sdf.parse(date);
+            sdf.applyPattern(AMZ_DATE_FORMAT_SHORT);
+            date = sdf.format(d);
+        } catch (ParseException e) {
+            throw new RuntimeException(e);
+        }
+        stringToSign.append(getScope(date, service)).append("\n");
+
+        // get hashedCanonicalRequest
+        byte[] hash = hash256(canonicalRequest);
+        String hashedRequest = hexEncode(hash);
+
+        stringToSign.append(hashedRequest);
+        return stringToSign.toString();
+    }
+
+    protected byte[] getSigningKey(String date, String service) {
+        return hmac(S3Constants.HMAC_SHA_256,
+                hmac(S3Constants.HMAC_SHA_256,
+                        hmac(S3Constants.HMAC_SHA_256,
+                                hmac(S3Constants.HMAC_SHA_256,
+                                        (S3Constants.AWS_V4 + s3Config.getSecretKey()).getBytes(StandardCharsets.UTF_8), date),
+                                S3Constants.AWS_DEFAULT_REGION),
+                        service),
+                S3Constants.AWS_V4_TERMINATOR
+        );
+    }
+
+    protected String getDate(Map<String, String> parameters, Map<String, List<Object>> headers) {
+        String date = null;
+        // if x-amz-date is specified, date should be the value
+        if (headers.containsKey(S3Constants.AMZ_DATE)) {
+            date = RestUtil.getFirstAsString(headers, S3Constants.AMZ_DATE);
+            return date;
+        }
+        // use Date header
+        date = RestUtil.getFirstAsString(headers, RestUtil.HEADER_DATE);
+        if (date == null) {
+            // must have a date in the headers
+            date = RestUtil.getRequestDate(s3Config.getServerClockSkew());
+        }
+
+        // convert date
+        SimpleDateFormat sdf = new SimpleDateFormat(HEADER_DATE_FORMAT, Locale.US);
+        try {
+            Date d = sdf.parse(date);
+            sdf.applyPattern(AMZ_DATE_FORMAT);
+            return sdf.format(d);
+        } catch (ParseException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    protected String getShortDate(String date) {
+        // Date must be consistent with timestamp, so extract it
+        // from previous date time format instead of get current date
+        String shortDate = "";
+        SimpleDateFormat sdf = new SimpleDateFormat(AMZ_DATE_FORMAT, Locale.US);
```

Comment:
another SimpleDateFormat constructor - use `ThreadLocal`s or switch to Java 8 `DateTime`

Reply:
Used DateTimeFormatter to replace SimpleDateFormat